models/source_list.py

```diff
@@ -53,10 +67,106 @@ class SourceList(Base, FileOnDiskMixin):
+        for key, value in kwargs.items():
+            if hasattr(self, key):
+                setattr(self, key, value)
+
+    @orm.reconstructor
+    def init_on_load(self):
+        Base.init_on_load(self)
+        FileOnDiskMixin.init_on_load(self)
+        self._data = None
+
+    def __repr__(self):
+        output = (
+            f'<SourceList(id={self.id}, '
+            f'image_id={self.image_id}, '
+            f'is_sub={self.is_sub}), '
+            f'num_sources= {self.num_sources}>'
+        )
+
+        return output
+
+    @property
+    def data(self):
+        """
+        The data in this source list. A table of sources and their properties.
+        """
+        if self._data is None and self.filepath is not None:
+            self._data = self.load()
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        # TODO: add astropy table?
+        if isinstance(value, pd.DataFrame):
+            value = value.to_records(index=False)
+
+        if not isinstance(value, np.ndarray) or value.dtype.names is None:
+            raise TypeError("data must be a pandas DataFrame or numpy recarray")
+
+        self._data = value
+        self.num_sources = len(value)
+
+    @property
+    def mjd(self):
+        """
+        The MJD of the image this source list was generated from.
+        """
+        if self.image is None:
+            return None
+        return self.image.mjd
+
+    def load(self):
+        """
+        Load this source list from the database.
+        """
+        # TODO: should we replace this with FITS and astropy tables?
+        return np.load(self.get_fullpath())  # this should always be a single file, right?
+
     def save(self):
         """
-        Save this source list to the database.
+        Save the data table to a file on disk.
         """
-        # TODO: Must implement this at some point!
-        self.filepath = uuid.uuid4().hex
+        if self.image is None:
+            raise ValueError("Cannot save source list without an image")
+
+        if self.data is None:
+            raise ValueError("Cannot save source list without data")
+
+        filename = self.invent_filename()
+
+        if self.is_sub:
+            filename += '_detections'
+        else:
+            filename += '_sources'
+
+        # TODO: use FITS and astropy tables instead?
+        filename += '.npy'
+        fullname = os.path.join(self.local_path, filename)
+        self.safe_mkdir(os.path.dirname(fullname))
+        np.save(fullname, self.data)
+
+        self.filepath = filename
+
 
+# add "property" attributes to SourceList referencing the image for convenience
+for att in [
+    'section_id',
+    'mjd',
+    'ra',
+    'dec',
+    'filter',
+    'filter_short',
+    'telescope',
+    'instrument',
+    'instrument_object',
+]:
+    setattr(SourceList, att, property(fget=lambda self, att=att: getattr(self.image, att) if self.image is not None else None))
```

Comment:
You've already explicitly defined @property mjd above

Reply:
oops!